vyper/semantics/analysis/local.py

```diff
@@ -790,3 +726,107 @@ def visit_IfExp(self, node: vy_ast.IfExp, typ: VyperType) -> None:
         self.visit(node.body, typ)
         validate_expected_type(node.orelse, typ)
         self.visit(node.orelse, typ)
+
+
+def _analyse_range_call(node: vy_ast.Call) -> list[VyperType]:
+    """
+    Check that the arguments to a range() call are valid.
+    :param node: call to range()
+    :return: None
+    """
+    validate_call_args(node, (1, 2), kwargs=["bound"])
+    kwargs = {s.arg: s.value for s in node.keywords or []}
+
+    start, end = (vy_ast.Int(value=0), node.args[0]) if len(node.args) == 1 else node.args
+
+    if "bound" in kwargs:
+        return _analyse_range_bound(start, end, kwargs["bound"], node)
+    if isinstance(start, vy_ast.Num):
+        return _analyse_range_constant(start, end, node)
+    return _analyse_range_sum(start, end, node)
+
+
+def _analyse_range_constant(
+    start: vy_ast.Num, end: vy_ast.VyperNode, node: vy_ast.Call
+) -> list[VyperType]:
+    """
+    Check that the arguments to a range(N, M) call are valid.
+    :param start: first argument to range()
+    :param end: second argument to range()
+    :param node: range() call node
+    :return: The common types of the arguments
+    """
+    if not isinstance(end, vy_ast.Num):
+        raise StateAccessViolation("Value must be a literal integer", end)
+    if end.value <= start.value:
+        raise StructureException("End must be greater than start", end)
+    return _get_common_argument_types(node)
+
+
+def _analyse_range_bound(
+    start: vy_ast.VyperNode, end: vy_ast.VyperNode, bound: vy_ast.VyperNode, node: vy_ast.Call
+) -> list[VyperType]:
+    """
+    Check that the arguments to a range(x, bound=N) or range(x, y, bound=N) call are valid.
+    :param start: first argument to range()
+    :param end: second argument to range()
+    :param bound: bound keyword argument to range()
+    :param node: range() call node
+    """
+    if not isinstance(bound, vy_ast.Num):
+        raise StateAccessViolation("Bound must be a literal", bound)
+    if bound.value <= 0:
+        raise StructureException("Bound must be at least 1", bound)
+    if (
+        isinstance(start, vy_ast.Num)
```

Comment:
not necessary for this PR but it might be nice if Num nodes had some function like `int_val()` -- this would make it so we don't have to do `isinstance` checks (e.g. `isinstance(bound, vy_ast.Int)` or `isinstance(start.value, int)`

Reply:
here's an example of triggering a strange error message:

![Screenshot from 2023-12-13 07-58-55](https://github.com/vyperlang/vyper/assets/3867501/0a36a9f2-98c3-4028-94e4-5e012ca57b1d)

(you can't really get a compiler panic or anything because of the check to `get_common_types()` later)

Reply:
yeah I think most isinstance calls can be avoided by using properties and duck typing
We can easily improve the error message here if you prefer

Reply:
i'm going to declare the `int_val()` thing as "out-of-scope" here :)

however, it did just occur to me that maybe for this particular error message, we should just outright ban the use of `bound=` when the arguments are literals?